flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java

```diff
@@ -159,16 +185,18 @@ private void checkCrashLoopBackoff(FlinkDeployment flinkApp, Configuration effec
         }
     }
 
-    protected boolean isClusterReady(FlinkDeployment dep) {
+    protected boolean isJmDeploymentReady(FlinkDeployment dep) {
         return dep.getStatus().getJobManagerDeploymentStatus() == JobManagerDeploymentStatus.READY;
     }
 
     protected void clearErrorsIfJobManagerDeploymentNotInErrorStatus(FlinkDeployment dep) {
-        if (dep.getStatus().getJobManagerDeploymentStatus() != JobManagerDeploymentStatus.ERROR) {
-            final ReconciliationStatus reconciliationStatus =
-                    dep.getStatus().getReconciliationStatus();
-            reconciliationStatus.setSuccess(true);
-            reconciliationStatus.setError(null);
+        FlinkDeploymentStatus status = dep.getStatus();
+        ReconciliationStatus reconciliationStatus = status.getReconciliationStatus();
+        if (status.getJobManagerDeploymentStatus() != JobManagerDeploymentStatus.ERROR
+                && reconciliationStatus
+                        .getLastReconciledSpec()
+                        .equals(reconciliationStatus.getLastStableSpec())) {
```

Comment:
nit: looks like we can directly use `reconciliationStatus.isLastReconciledSpecStable`.

Besides, I want to make sure I under the new condition correctly: previously, we will try to clear errors and once the JM deployment is not in ERROR status, we do the cleaning. After introducing the `isLastReconciledSpecStable` check, we only `markReconciledSpecAsStable` when `isJmDeploymentReady` && `observeFlinkCluster` && `reconciliationStatus.getState() != ReconciliationStatus.State.ROLLED_BACK`. As a result, if JM is in states like `DEPLOYED_NOT_READY`, the error in status will not be cleared.  
It seems that the later clear will not influence the correctness of the current logic as whether error info is null is only checked in `isJobUpgradeInProgress` but may make the method name `clearErrorsIfJobManagerDeploymentNotInErrorStatus` misleading. WDYT?

Reply:
The intention here was to avoid losing error information that caused a rollback operation thats why I introduced the additional isStable check. I have now updated the method name to: `clearErrorsIfDeploymentIsHealthy` that is more appropriate